toolchain/semantics/semantics_handle_paren.cpp

```diff
@@ -19,17 +25,36 @@ auto SemanticsHandleParenExpression(SemanticsContext& context,
 auto SemanticsHandleParenExpressionOrTupleLiteralStart(
     SemanticsContext& context, ParseTree::Node parse_node) -> bool {
   context.node_stack().Push(parse_node);
+  context.ParamOrArgStart();
+  return true;
+}
+
+auto SemanticsHandleTupleLiteralComma(SemanticsContext& context,
+                                      ParseTree::Node /*parse_node*/) -> bool {
+  context.ParamOrArgComma(/*for_args=*/true);
   return true;
 }
 
 auto SemanticsHandleTupleLiteral(SemanticsContext& context,
                                  ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleTupleLiteral");
-}
+  auto refs_id = context.ParamOrArgEnd(
+      /*for_args=*/true, ParseNodeKind::ParenExpressionOrTupleLiteralStart);
 
-auto SemanticsHandleTupleLiteralComma(SemanticsContext& context,
-                                      ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleTupleLiteralComma");
+  context.node_stack()
+      .PopAndDiscardSoloParseNode<
+          ParseNodeKind::ParenExpressionOrTupleLiteralStart>();
+  const auto& node_block = context.semantics_ir().GetNodeBlock(refs_id);
+  llvm::SmallVector<SemanticsTypeId> type_ids;
+  type_ids.reserve(node_block.size());
+  for (auto node : node_block) {
+    type_ids.push_back(context.semantics_ir().GetNode(node).type_id());
```

Comment:
I think types need to be canonicalized here otherwise it may add invalid type ids to the tuple type block [here](https://github.com/carbon-language/carbon-lang/pull/2992/files#diff-0aac17c90333793c50c1315abd8f2c66f6d3a6a4ca2c98a6bfee8083afce2afeR659) @jonmeow could you please let me know your thought regarding this.

Reply:
Put differently, an invalid type id here indicates a bug, not a need for canonicalization.